sql+storage: subsource dependency inversion v2

sploiselle opened the pull request:
### Motivation

This PR adds a known-desirable feature. Progresses #24773
This PR fixes a known bug. Fixes #26465

### Tips for reviewer

This PR un-inverts subsource dependencies (for data-bearing subsources; progress subsources remain untouched). The gist of how we do this is whenever we add a subsource to a source:
1. Modify the `IngestionDescription`'s `SourceDesc` to contain a reference to the subsource (e.g. adding it to the PG publication tables and generating its table casts)
2. Modify the ingestion's `source_exports` to contain the association betwen with `GlobalId` and the `output_index`.

This is essentially the same process we underwent before, but we now do this in response to a `create_collection` whose `DataSource` is `DataSource::IngestionExport` (a new enum which describes subsources).

If we are adding a subsource via `ALTER SOURCE...ADD SUBSOURCE`, we will also reconcile the source's state such that it contains references only to tables referenced by existing subsources.

When dropping subsources, we will now just drop the source via `DROP SOURCE` (i.e. `ALTER SOURCE...DROP SUBSOURCE` goes away). When dropping sources, we detect if they're `DataSource::IngestionExport` and if so, we remove the `source_export` and re-render the ingestion.

This has the slightly annoying side effect that we have to do a bunch of work to produce roundtrippable `SHOW CREATE SOURCE` statements, but all of that work is less code than `ALTER SOURCE...DROP SUBSOURCE` was, so it's a win on the balance.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Let users remove subsources from PostgreSQL and MySQL sources using `DROP SOURCE`. We previously required the command `ALTER SOURCE...DROP SUBSOURCE` or `ALTER SOURCE...DROP TABLE`; however, those commands have been deprecated in favor of `DROP SOURCE`.
  - Require users to specify `CASCADE` to drop PostgreSQL and MySQL sources with active subsources (assuming they want to drop the subsources, as well). Previously, dropping a PostgreSQL or MySQL source automatically dropped all of its subsources.
  - Allow `ALTER...OWNER` on a primary source's subsources, i.e. a source and its subsources may now have different owners.

sploiselle commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -4125,6 +4125,8 @@ pub fn plan_drop_objects(
     }
 
     let drop_ids = scx.catalog.object_dependents(&referenced_ids);
+    tracing::info!("drop_ids {:?}", drop_ids);
```
removed locally.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -618,298 +679,116 @@ where
-            // Assess dependency since, which we have to fast-forward this
-            // collection's since to.
-            let storage_dependencies = collection.description.get_storage_dependencies();
+            tracing::info!(
```
Removing locally.

rjobanp commented on src/sql/src/pure.rs:
```diff
@@ -1310,29 +1284,27 @@ async fn purify_create_source(
 /// we are permitted to use async code.
```
docstring needs updating w.r.t return types

sploiselle replied:
ty

rjobanp commented on src/sql/src/pure.rs:
```diff
@@ -1310,29 +1284,27 @@ async fn purify_create_source(
 
     // Get connection
-    let pg_source_connection = {
+    let (source_name, pg_source_connection, mut current_references) = {
```
nit: there's a baked-in assumption here that this has to be a PG source -- despite the `purify_alter_source` method potentially applying to any source that has subsources

I can refactor this when I implement alter source for mysql, but since we're already changing a lot here it could be nice to rename `pg_source_connection` to `source_connection` and make things more generic below until we get to the pg-specific stuff around replication-slots, publications, etc (similar to `purify_create_source`)

rjobanp commented on src/sql/src/pure.rs:
```diff
@@ -1406,48 +1400,34 @@ async fn purify_alter_source(
-            native_idx,
-        );
-    }
+    let unresolved_source_name = UnresolvedItemName::from(source_name.full_item_name().clone());
```
how is this different from `outer_source_name` above? Can you leave some comments explaining what's going on with the name mangling?

sploiselle replied:
These should have had better variable names--they're different types. One is a `ResolvedItemName`, one is an `UnresolvedItemName`. Amended locally and will push that up in a bit.

rjobanp reviewed (commented):
I just got through the first bits of the change (SourceExport structure, purification of create / alter source) - I like how much simpler a lot of the purification logic is!

rjobanp commented on src/storage-controller/src/lib.rs:
```diff
@@ -892,10 +951,9 @@ where
-            )?;
+            for id in new_source_exports {
+                self.install_dependency_read_holds([id].into_iter(), &storage_dependencies)?;
+            }
```
outdated based on an early commit

ParkMyCar commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -3161,77 +3389,49 @@ impl Coordinator {
         };
 
         match action {
+            // TODO: Dropping a subsource via `ALTER SOURCE` only removes the
+            // references to the object
```
What would it ideally also do?

sploiselle replied:
Ah, this is actually done in this PR and should have been removed.

ParkMyCar commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -3283,7 +3513,7 @@ impl Coordinator {
                 let plan = match mz_sql::plan::plan(
```
Just about this we mark the ID of the `cur_entry` as unresolvable for planning, remind me again why we do that?

sploiselle replied:
Clarified this in a call. The issue here is that we need to re-plan the statement and want to ensure we don't get an "object with that name already exists" error.

ParkMyCar commented on src/catalog/protos/objects.proto:
```diff
@@ -664,13 +664,23 @@ message AuditLogEventV1 {
     optional string new_history = 3;
   }
 
+  message ToNewIdV1 {
```
I was thinking about this a bit more, what if we re-wrote the entire audit log with the new IDs?

sploiselle replied:
Chatted about this during a call and I have a lot of apprehensions about that because it would require we update _all_ data system-wide with the new IDs to retain the data's referential integrity.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -270,82 +476,109 @@ impl Coordinator {
+                            id,
+                            external_reference,
+                        } => {
+                            // Propagate source's compaction window if this
+                            // subsource does not have its own value specified.
```
We re-set these during bootstrapping, but it doesn't look like the same logic gets applied. cc @mjibson

sploiselle replied:
cc @maddyblue -- any idea what we ought to do here, or have you moved onto other projects?

maddyblue replied:
Reading over https://github.com/MaterializeInc/materialize/pull/24010/files it looks like on restart subsource retain history does not inherit from the source. #26650 was already tracking another cause of this (so subsources were doubly broken here). I think this PR should not worry about that because it's already big enough and doesn't work anyway. I'll fix it later.

sploiselle replied:
Thank you!

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -94,6 +95,11 @@ pub enum DataSource {
     Webhook,
+    /// This source receives its data from another source's export.
+    SourceExport {
+        id: GlobalId,
```
Ah, yeah––good call. Will rename all of this stuff.

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -734,6 +743,10 @@ impl<T: Timestamp> CollectionState<T> {
             DataSource::Webhook
             | DataSource::Introspection(_)
             | DataSource::Other(_)
+            // This isn't quite right because a source export runs on the
+            // ingestion's cluster, but we don't have the ability to perform
+            // cross-referenced lookups here (at least not yet).
```
This is how things currently work and once this change lands, I'll change this. This is tracked in #24235; adding a link here.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -530,61 +532,115 @@ where
             .await?;
 
-        let mut to_create = Vec::with_capacity(to_register.len());
+        // Reorder in dependency order.
```
Yes

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -649,6 +706,7 @@ where
                         &storage_dependencies,
                     )?;
                 }
+                DataSource::SourceExport { .. } => todo!(),
```
If you review commit by commit, need to check that `todo`s don't make it into the final commit. We have a CI lint that ensures no `todo!`s make it into `main`.

sploiselle commented on src/catalog/src/memory/objects.rs:
```diff
@@ -388,6 +388,11 @@ pub enum DataSourceDesc {
     Ingestion(IngestionDescription<(), ReferencedConnection>),
     /// Receives data from some other source
```
There's a `TODO` explaining that this should just be removed. Might have been added in a subsequent commit.

sploiselle commented on src/catalog/src/memory/objects.rs:
```diff
@@ -547,6 +556,9 @@ impl Source {
                 },
                 None => (None, None),
             },
+            // This isn't quite right because these values are encoded--just
+            // that they're encoded by the source.
+            DataSourceDesc::SourceExport { .. } => (None, None),
```
Ah, this comment lacks all of the context necessary to be meaningful; will remove it. In short, PG and MySQL sources have a kind of bespoke format that's unspecifiable in the the syntax.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -146,13 +146,348 @@ pub(crate) async fn migrate(
+    let mut remaining_updates = VecDeque::from_iter(needs_new_id.iter().cloned());
+    // If this item's ID must be moved forward, so too must every item that
+    // depends on it except for its primary source ID.
+    while let Some(id) = remaining_updates.pop_front() {
+        needs_new_id.push(id);
+        remaining_updates.extend(
+            conn_catalog
+                .state()
+                .get_entry(&id)
+                .used_by()
+                .iter()
+                .filter(|id| !source_ids_of_updated_subsources.contains(id))
+                .cloned(),
+        );
+    }
```
Cleared this up in the chat--swapping `GlobalId`s would introduce some complex data anomalies.

sploiselle commented on src/adapter/src/catalog/open.rs:
```diff
@@ -1189,6 +1189,10 @@ impl Catalog {
     ) -> Result<CatalogState, Error> {
         let mut state = state.clone();
+
+        // TODO: after we migrate subsources to the new dependency structure (in
+        // v0.96, so in v0.97 forward), we can refactor this to assume that
+        // items are topologically sorted by their `GlobalId`.
```
This gets run before the durable migration. Anything that runs _after_ the durable migration can assume that the dependency structure can be inferred by the `GlobalId`'s `Ord` implementation, which is why this code can't be fixed just yet (i.e. we need to open the catalog with the inverted dependency structure).

sploiselle commented on src/adapter/src/catalog/state.rs:
```diff
@@ -489,13 +489,17 @@ impl CatalogState {
-                dependents.extend_from_slice(&self.item_dependents(subsource_id, seen));
-            }
             dependents.push(object_id);
+            // We treat the progress collection as if it depends on the source
+            // for dropping. We have additional code in planning to create a
+            // kind of special-case "CASCADE" for this dependency.
+            if let Some(progress_id) = entry.progress_id() {
+                dependents.extend_from_slice(&self.item_dependents(progress_id, seen));
+            }
```
Unfortunately we can't "fix" this because it's in the `uses` set. I'm doing this kind of annoyingly because dropping the source should also drop the progress collection (even though this is backward from a dependency perspective). This is essentially just a stopgap until we get rid of the progress collections completely (they'll just become the source's data, and Kafka sources will start outputting a "subsource," though that subsource will be called a "table").

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -1158,41 +1037,80 @@ where
+
+            to_drop.insert(id);
+
+            // If we are dropping source exports, we need to the ingestion that
+            // it runs on.
```
Modify! ty for the catch.

sploiselle commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -508,7 +508,7 @@ impl Coordinator {
             }
             if !sources_to_drop.is_empty() {
-                self.drop_sources(sources_to_drop);
+                self.drop_sources(sources_to_drop).await;
```
This is just "infected" with async because executing some sources requires async (though this does not). Will amend with a comment that this needs to have the same behavior as `create_collections`, which doesn't run in a task.

sploiselle commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -467,99 +466,49 @@ impl Coordinator {
+            Statement::AlterSource(_) | Statement::CreateSink(_) => {
+                // These must have only only a single statement.
+                let stmt = stmts.remove(0);
+                assert!(stmts.is_empty());
```
We chatted about this and I'll add a wrapping struct to these values that has a little more structure.

sploiselle replied:
Totally refactored how purification is handled––it's now type safe within an inch of its life.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -256,6 +258,210 @@ impl Coordinator {
+    }
+
+    /// Prepares a `CREATE SOURCE` statement to create its progress subsource,
+    /// the primary source, and any data-bearing subsources.'
```
Added

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -3244,9 +3444,6 @@ impl Coordinator {
                                         WithOptionValue::UnresolvedItemName(
                                             column_qualified_reference,
                                         ) => {
-                                            mz_ore::soft_assert_eq_or_log!(
-                                                column_qualified_reference.0.len(), 4
-                                            );
```
This already issued a warn a few lines below, so figured that sufficed.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -146,13 +146,348 @@ pub(crate) async fn migrate(
+///   `GlobalId`s they wish to reassign to `needs_new_id`.
+/// - Assumes all items in `needs_new_id` are present in `conn_catalog` and that
+///   their types do not change.
+fn assign_new_user_global_ids(
```
This was so clutch--I had no idea this existed. ty!

sploiselle commented:
@rjobanp @ParkMyCar Implemented everything y'all suggested.

- Renaming `DataSource::SourceExport` to `DataSource::IngestionExport` with an `ingestion_id` field didn't absorb neatly, so I just refactored it in a final commit. I don't love it but I also didn't want to spend 2 hours pushing it back through all of the commits. If that bugs anyone I am happy to let them do the work. [storage: rename controller::DataSource::SourceExport to IngestionExport](https://github.com/MaterializeInc/materialize/pull/26556/commits/abc14cd209a6642cbbd3b3de3af6dfb3c3a7fa92)
- Refactored the values we expect from purification. I believe this is as type safe as you could reasonably expect.
[adapter: refactor structure of handling purified statements](https://github.com/MaterializeInc/materialize/pull/26556/commits/d362250b9f6ad9b95035cb623f7df83dbb88b54c)
[d362250](https://github.com/MaterializeInc/materialize/pull/26556/commits/d362250b9f6ad9b95035cb623f7df83dbb88b54c)

@ParkMyCar I amended the migration to update comment `GlobalId`s, as well. If I can get an explicit approval from you on everything that would be great. I won't merge until the storage team also concludes they're happy with the work.

@MaterializeInc/storage ready for another review. I am OOO for the next week, though, so no rush but I need the feedback in by April 19 so I can address and merge when I return.

petrosagg reviewed (commented):
Here is a summary of my review since I will be out for two weeks while this is being worked on.

I went over all the code and feel like I have a good grasp of what's going on for most of it. That said, the big changes in `lib.rs` in the controller and the the new way sources with subsources are planned is a little less clear to me. This is a very big PR and changes the code in fundamental ways so I would want to see some long text somewhere (PR description, a markdown doc, a comment) that explains how everything works. Something that goes over what happens to a source statement as it is typed by the user, goes through purification, gets planned, sequenced etc. There are a lot of transformations along the way that are only evident by intense studying of the code. The same request applies for an `ALTER SOURCE` statement.  We now have a `SET ADD SUBSOURCE` internal statement that it is still unclear to me what it does.

Beyond that, we definitely want to create a tracking issue that contains all the todo items that are to be performed after this gets released. I tried to mark the ones I saw with comments but I may have missed some. If we don't create a tracking issue we will probably forget/miss them.

Unfortunately I can't give an approved review at the current state of the PR as some of the comments need to be worked on. So if merging this while I'm away becomes a pressing priority it will have to be at Sean's discretion.

ParkMyCar commented on test/sqllogictest/object_ownership.slt:
```diff
@@ -2388,13 +2388,13 @@ SELECT mz_sources.name, mz_roles.name
+accounts       materialize
+auctions       materialize
+bids           materialize
+organizations  materialize
```
+1 I'm pretty sure this shouldn't change

sploiselle replied:
I believe the updated test in my commit reflects the behavior we expect. We can see that changing an owner does not change the ownership of the object's dependents using this example.

Log in as `mz_system`

```
psql -h localhost -U mz_system materialize -p 6877
```

Then:

```
SET CLUSTER TO default;
CREATE TABLE t (a int);
CREATE MATERIALIZED VIEW v AS SELECT sum(a) FROM t;
CREATE ROLE joe;

SELECT mz_tables.name AS table_name, mz_roles.name AS role_name
  FROM mz_tables
  LEFT JOIN mz_roles ON mz_tables.owner_id = mz_roles.id
  WHERE mz_tables.name = 't';

 table_name | role_name 
------------+-----------
 t          | mz_system

SELECT mz_materialized_views.name AS view_name, mz_roles.name AS role_name
  FROM mz_materialized_views
  LEFT JOIN mz_roles ON mz_materialized_views.owner_id = mz_roles.id
  WHERE mz_materialized_views.name = 'v';

 view_name | role_name 
-----------+-----------
 v         | mz_system


ALTER TABLE t OWNER TO joe;

SELECT mz_tables.name AS table_name, mz_roles.name AS role_name
  FROM mz_tables
  LEFT JOIN mz_roles ON mz_tables.owner_id = mz_roles.id
  WHERE mz_tables.name = 't';

 table_name | role_name 
------------+-----------
 t          | joe


SELECT mz_materialized_views.name AS view_name, mz_roles.name AS role_name
  FROM mz_materialized_views
  LEFT JOIN mz_roles ON mz_materialized_views.owner_id = mz_roles.id
  WHERE mz_materialized_views.name = 'v';

 view_name | role_name 
-----------+-----------
 v         | mz_system

```

`v` depends on `t`, but changing `t`'s owner doesn't affect `v`. Similarly, subsources depend on sources, but changing the source's owner doesn't affect the subsource.

@jkosh44 Can you validate that this behavior is correct?

In testing this, though, I have realized this PR prevents changing ownership of a subsource directly. I'll fix that in an additional commit and push that back up.

sploiselle replied:
@jkosh44 ping

jkosh44 replied:
Sorry for the delay in responding. It could go either way here. Changing the owner of a table does not affect the owner of any dependent views. However, changing the owner of a table does also change the owner of all indexes on that table. Similarly, changing the owner of cluster changes the owner of all replicas within that cluster. This behave usually matches the behavior of `DROP`. `DROP TABLE _` drops the table and all indexes without the need for `CASCADE`. Similarly, `DROP CLUSTER _` drops the cluster and all replicas without the need for `CASCADE`.

So if `DROP SOURCE _` works without `CASCADE` then I think the behavior of this PR is correct. Otherwise, we should revert to the old behavior.

sploiselle replied:
The [SQL council concluded](https://materializeinc.slack.com/archives/C063H5S7NKE/p1713344604109359) that dropping a source w/ subsources requiring `CASCADE` is acceptable, so going to leave the PR as is.

ParkMyCar reviewed (commented):
Can you add a SLT or TD test that creates a source with subsources and then selects from `mz_objects` and asserts their global ids? The goal being we can assert GlobalIds for subsources are indeed in sorted order

rjobanp commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+                .iter()
+                .map(|t| {
+                    UnresolvedItemName(vec![
+                        Ident::new_unchecked("mysql"),
```
they shouldn't actually be placed in a database named mysql in materialize, but to fit the model of a 3-layer catalog / UnresolvedItemName when referencing mysql subsources we do use the mysql placeholder db: https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/pure/mysql.rs#L30-L43

sploiselle commented on test/testdrive/mz-depends.td:
```diff
@@ -24,9 +24,9 @@ ALTER SYSTEM SET enable_connection_validation_syntax = true
-  JOIN mz_sources AS s ON s.id = d.referenced_object_id
-  JOIN mz_sources AS top_level_s ON top_level_s.id = d.object_id
-  WHERE top_level_s.name = 'with_subsources';
+  JOIN mz_sources AS s ON s.id = d.referenced_object_id OR s.id = d.object_id
+  JOIN mz_sources AS top_level_s ON top_level_s.id = d.object_id OR top_level_s.id = d.referenced_object_id
+  WHERE top_level_s.name = 'with_subsources' AND (s.type = 'progress' OR s.type = 'subsource');
```
Primary sources still depend on progress subsources, so there are dependencies in both directions. This will eventually be fixed by merging progress subsources and sources themselves in #20208

sploiselle commented on test/restart/mzcompose.py:
```diff
@@ -395,12 +395,13 @@ def workflow_bound_size_mz_status_history(c: Composition) -> None:
                 """
-                > ALTER CONNECTION kafka_conn SET (SECURITY PROTOCOL PLAINTEXT)
+                > ALTER CONNECTION kafka_conn SET (BROKER 'dne') WITH (VALIDATE = false);
+                > ALTER CONNECTION kafka_conn SET (BROKER '${testdrive.kafka-addr}') WITH (VALIDATE = true);
```
Yes––this test requires that we restart the source some number of times to generate events. This PR changes the way altering a connection works––it doesn't actually restart the ingestion if the description matches (because it doesn't meaningfully change anything, so restarting the ingestion is pointless).

sploiselle commented on test/pg-cdc/alter-source.td:
```diff
@@ -129,29 +129,25 @@ table_e               subsource
 contains:SOURCE "mz_source" is not a subsource of "mz_source"
 
+! DROP SOURCE table_f;
+contains:SOURCE "table_f" is a subsource which must be dropped using ALTER SOURCE...DROP SUBSOURCE
```
I'll file an issue to fix this after this PR merges. I propose we let users drop sources using `DROP SOURCE` unless they refer to a table referred to with `TEXT COLUMNS`. If we allow sources with `TEXT COLUMN` references to be dropped, we'll have dangling references to `TEXT COLUMNS` that are no longer in use and we do not currently have logic to garbage collect those references. We _can_ do that in the future, but I tried to minimize the complexity here.

sploiselle replied:
I removed `ALTER SOURCE...DROP SUBSOURCE`, so going to mark this as resolved.

sploiselle commented on src/storage/src/source/generator.rs:
```diff
@@ -44,30 +44,39 @@ enum GeneratorKind {
         generator: Box<dyn Generator>,
         tick_micros: Option<u64>,
+        // Load generators cannot be rendered until all of their exports are
+        // present.
```
We chatted about this in real time during our pre-review of this PR. At that time, you said that this was unfortunate and should be fixed eventually. I will file an issue that these need to be re-designed.

> What happens if render the load gen with fewer subsources?

We don't support that. All load generator sources require `FOR ALL TABLES`. If we did support producing a subset of tables, we either would have to have not tested the scenario (somewhat unlikely) or it would have failed because we never would have produced data. Fortunately, we can see that we just [don't support rendering a subset of the subsources](https://github.com/MaterializeInc/materialize/blob/4b0f90f590dc0b0824e6a14935160a3f2253e441/src/sql/src/pure.rs#L1159-L1179).

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -2317,9 +2334,11 @@ impl_display_t!(AlterSourceAddSubsourceOption);
     ResetOptions(Vec<CreateSourceOptionName>),
+    // This is meant for internal use only and cannot be roundtripped through
+    // parsing.
+    SetAddSubsourceOptions(Vec<AlterSourceAddSubsourceOption<T>>),
```
Removed this, though I don't think adding AST elements inaccessible to the user is so controversial.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+                    ])
+                })
+                .collect(),
+            _ => unreachable!("only PG and MySQL sources have non-self source exports"),
```
Big derp on my end. Thank you.

sploiselle replied:
Fixed this on this branch so marking as resolved.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+                Statement::CreateSubsource(create_subsource_stmt) => {
+                    create_subsource_stmt.of_source = Some(RawItemName::Name(source_name.clone()));
+
+                    let pos = create_subsource_stmt
+                        .with_options
+                        .iter()
+                        .position(|o| o.name == CreateSubsourceOptionName::References)
+                        .expect("subsources must be of references type");
+
+                    create_subsource_stmt.with_options.swap_remove(pos);
```
Yes! No idea why I did this.

sploiselle replied:
Ah, I had originally used the returned value, but moved to just using a new option name entirely. Anyway––fixed locally and will be in next push.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+/// This function also adjusts items' `GlobalId`s such that the dependency graph
+/// can be inferred using the ordering of `GlobalId`s, i.e. all of an item's
+/// dependencies have `GlobalId`s less than its own.
+fn subsource_rewrite_v0_96(
```
I added a few but will add more.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+    // If this item's ID must be moved forward, so too must every item that
+    // depends on it except for its primary source ID.
+    while let Some(id) = remaining_updates.pop_front() {
+        needs_new_id.push(id);
```
Thank you for catching this. I fixed it by reverse iterating over `needs_new_id` and only keeping the first entry (i.e. the last entry) which ensures that each item is in a greater position than all of its dependencies.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+                            *id = new_id.to_string();
+                        }
+                    }
+                    Err(e) => {
```
Only if there is corruption

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+    let update_items: Vec<_> = tx
+        .get_items()
+        .into_iter()
+        .filter_map(|item| match needs_new_id.contains(&item.id) {
```
Yeah––I am aware this is quadratic but it's O([number of items in catalog] * [number of updated item]) and the number of updated items should be small. Certainly no harm in putting it in a hashset, though.

sploiselle replied:
Put these values in a set.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+
+    let mut new_id_mapping = BTreeMap::new();
+
+    for (item, new_id) in update_items.into_iter().zip_eq(new_ids.into_iter()) {
```
Thank you for catching this. I'll make sure someone else pores over this before we merge.

sploiselle commented on src/storage-types/src/sources.rs:
```diff
@@ -77,6 +78,12 @@ pub struct IngestionDescription<S: 'static = (), C: ConnectionAccess = InlinedCo
     ///
     /// Note that this does _not_ include the remap collection, which is tracked
     /// in its own field.
+    // TODO: after implementing subsource dependency inversion, we should
```
Yes.

sploiselle replied:
I moved all of these into GH tracking issues.

sploiselle commented on src/storage-types/src/sources/load_generator.rs:
```diff
@@ -119,6 +120,14 @@ impl SourceConnection for LoadGeneratorSourceConnection {
+        self.load_generator
+            .views()
+            .iter()
+            .position(|(view_name, _)| view_name == &name.0[2].as_str())
```
Good call.

sploiselle commented on src/storage-types/src/sources/mysql.rs:
```diff
@@ -249,6 +253,21 @@ pub struct MySqlSourceDetails {
+            .iter()
+            .position(|t| {
+                let inner = &name.0;
+                t.schema_name == inner[1].as_str() && t.name == inner[2].as_str()
```
Done locally.

sploiselle commented on src/storage-types/src/sources/mysql.rs:
```diff
@@ -165,6 +165,10 @@ impl<C: ConnectionAccess> SourceConnection for MySqlSourceConnection<C> {
         vec![]
     }
+
+    fn output_idx_for_name(&self, name: &mz_sql_parser::ast::UnresolvedItemName) -> Option<usize> {
```
Done locally.

sploiselle commented on src/storage-types/src/sources/postgres.rs:
```diff
@@ -282,6 +287,24 @@ pub struct PostgresSourcePublicationDetails {
+    ) -> Option<usize> {
+        // We do not validate the database qualification because we don't have
+        // it stored anywhere on the source itself, so we assume that the
+        // database is correctly validated during purification.
```
Done locally.

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -109,6 +115,8 @@ pub enum DataSourceOther {
     /// Compute maintains, i.e. it is a `MATERIALIZED VIEW`.
     Compute,
     /// Some other sources writes this data, i.e. a subsource.
+    // TODO: this can be removed in the version after we invert subsource
```
I'm going to close all of these out because I will file a tracking issue.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -256,6 +257,179 @@ impl Coordinator {
 
+    /// Subsources are planned differently from other statements because they
+    /// are typically synthesized from other statements, e.g. `CREATE SOURCE`.
+    /// Because of this, we have usually "missed" the opportunity to plan them
```
It's usually because bootstrapping uses the normal `plan_statement` path. Clarified.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -256,6 +257,179 @@ impl Coordinator {
+        let mut create_source_plans = Vec::with_capacity(create_subsource_stmts.len() + 2);
+        let mut of_source = None;
+
+        for stmt in [
+            Statement::CreateSubsource(create_progress_subsource_stmt),
+            Statement::CreateSource(create_source_stmt),
+        ]
+        .into_iter()
+        .chain(
+            create_subsource_stmts
+                .into_iter()
+                .map(Statement::CreateSubsource),
+        ) {
```
Sure thing. Done locally.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -256,6 +257,179 @@ impl Coordinator {
+                        create_source_plans.len(),
+                        1,
+                        "must create progress subsource before primary source"
+                    );
```
Sure thing. Done locally.

sploiselle commented on src/catalog/src/memory/objects.rs:
```diff
@@ -617,12 +634,25 @@ impl Source {
             }
+            //  DataSourceDesc::Source represents subsources, which are
+            //  accounted for in their primary source's ingestion. However,
+            //  maybe we can fix this.
```
I fixed this comment by removing the implicit TODO.

sploiselle commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -4114,13 +4124,70 @@ pub fn plan_drop_objects(
-    Ok(Plan::DropObjects(DropObjectsPlan {
+    // Check that dropping this object does not have any implicit dependencies
+    // (e.g. `TEXT COLUMNS` references for subsources).
+    for id in &referenced_ids {
```
Right now, we ensure that all of the `TEXT COLUMNS` values refer to columns present in the publication tables. 

Adding subsources now drops tables from the publication if they are no longer referred to.

To prevent `TEXT COLUMNS` from referring to tables that we remove from the publication, I figured this was the easiest way to keep things in sync. Ultimately, we should just remove `ALTER SOURCE...DROP SUBSOURCE` entirely and reconcile away dangling text column references when adding subsources. To minimize the complexity of the diff, I just added this rather than removing the command entirely.

However, explaining all of this makes me realize that this stop gap is more confusing than just fixing the issue. I'll remove this code, alongside `ALTER SOURCE...DROP SUBSOURCE`. I'll change the code such that adding a subsource reconciles the `TEXT COLUMNS` references. I'll also file a `TODO` to allow ad-hoc reconciliation of the `TEXT COLUMNS` references without adding a subsource.

sploiselle replied:
I resolved this by removing the code tagged here, deprecated `ALTER SOURCE...DROP SUSBOURCE`, and fixed `SHOW CREATE SOURCE` so that it always provides roundtrip-able values. We might be able to simplify the `SHOW CREATE SOURCE` implementation if we move `TEXT COLUMNS` to the `CREATE SUBSOURCE` statement, which I filed #26774 to track.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +155,383 @@ pub(crate) async fn migrate(
+        }
+
+        item.create_sql = stmt.to_ast_string_stable();
+        tx.update_item(id, item).expect("txn must succeed");
```
Oh nice––had no idea. Fixed locally and will be visible on next push.

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -1098,6 +1099,9 @@ impl AstDisplay for CreateSubsourceOptionName {
                 f.write_str("REFERENCES");
             }
+            CreateSubsourceOptionName::ExternalReference => {
+                f.write_str("EXTERNAL REFERENCE");
```
Sure thing. The `CREATE SUBSOURCE` statement is something invisible and inaccessible to users, so users simply don't interact with it. This was also discussed [in the SQL council slack channel](https://materializeinc.slack.com/archives/C063H5S7NKE/p1712753147530439?thread_ts=1712751579.916709&cid=C063H5S7NKE). I'll make sure to include more details about this in the design doc that I'll add on the next push.

sploiselle commented on src/sql-parser/tests/testdata/create:
```diff
@@ -655,4 +655,4 @@ CREATE SUBSOURCE IF NOT EXISTS source.sub (a int, b text) WITH (PROGRESS, REFERE
 ----
 CREATE SUBSOURCE IF NOT EXISTS source.sub (a int4, b text) WITH (PROGRESS, REFERENCES)
```
There is no new user-facing SQL.

sploiselle replied:
Done; ty for the nudge

sploiselle commented on src/sql/src/plan/error.rs:
```diff
@@ -341,6 +338,9 @@ impl PlanError {
                 "subsources referencing table: {}",
                 itertools::join(target_names, ", ")
             )),
+            Self::DropSourceUnavailableForSubsource { source, subsource } => {
```
This error got removed, but ty for the reminder!

sploiselle commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1623,22 +1620,33 @@ pub fn plan_create_subsource(
+    } else if references {
+        DataSourceDesc::Source
+    } else {
+        panic!("must have specified one")
```
Done locally

sploiselle commented on src/sql/src/plan/statement/show.rs:
```diff
@@ -391,14 +391,14 @@ fn show_subsources<'a>(
-            JOIN mz_internal.mz_object_dependencies deps ON subsources.id = deps.referenced_object_id
-            JOIN mz_sources AS sources ON sources.id = deps.object_id
-        WHERE {}",
+            JOIN mz_internal.mz_object_dependencies deps ON (subsources.id = deps.object_id OR subsources.id = deps.referenced_object_id)
```
We will need to do this until we merge the notion of progress subsources and sources themselves (primary sources depend on the progress subsource and data-bearing subsources depend on the primary source).

sploiselle commented on test/testdrive/statistics-deletion.td:
```diff
@@ -139,10 +139,11 @@ WHERE s.name IN ('bids') AND sch.name = 'subsources_cluster'
-> DROP SOURCE subsources_cluster.s
+> DROP SOURCE upsert_size CASCADE
+> DROP SOURCE upsert_cluster CASCADE
+# todo?
```
Fixed this.

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -256,6 +257,179 @@ impl Coordinator {
+        };
+        Ok(CreateSourcePlans {
+            source_id,
+            plan,
```
Done locally and will be included in next push.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -3103,184 +2998,82 @@ where
-            // Fill in the storage dependencies.
-            let collection = self.collection_mut(id).expect("known to exist");
+        if PartialOrder::less_than(&collection.implied_capability, &dependency_since) {
+            // Patch up the implied capability + maybe the persist shard's
```
restored the substance of the prior comment.

sploiselle commented on test/restart/mzcompose.py:
```diff
@@ -107,13 +107,13 @@ def workflow_github_17578(c: Composition) -> None:
-              JOIN mz_sources AS top_level_s ON top_level_s.id = d.object_id
-              WHERE top_level_s.name = 'with_subsources';
+              JOIN mz_sources AS s ON s.id = d.referenced_object_id OR s.id = d.object_id
+              JOIN mz_sources AS top_level_s ON top_level_s.id = d.object_id OR top_level_s.id = d.referenced_object_id
```
This is the same as above--primary sources still depend on progress subsources.

rjobanp commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +156,482 @@ pub(crate) async fn migrate(
+    // because individually updating items can perform poorly.
+    let mut updated_items = BTreeMap::new();
+
+    // Get all sources in this TXN.
+    let mut source_map: BTreeMap<_, _> = tx
+        .get_items()
+        .into_iter()
+        .filter(|item| conn_catalog.get_item(&item.id).item_type() == CatalogItemType::Source)
+        .map(|item| (item.id, item))
+        .collect();
+
+    // Iterate over all sources in the catalog.
```
What triggers sources to be already in the Transaction? Down below you have a
```source_map
    .remove(&primary_source_id)
    .expect("source exists");
```
which makes it seem that all sources are part of the transaction? Why do we need to do this list/filter operation to find sources on both the txn and the catalog?

rjobanp reviewed (commented):
A few comments on the adapter changes -- will go through the rest tomorrow

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +156,482 @@ pub(crate) async fn migrate(
+    // because individually updating items can perform poorly.
+    let mut updated_items = BTreeMap::new();
+
+    // Get all sources in this TXN.
+    let mut source_map: BTreeMap<_, _> = tx
+        .get_items()
+        .into_iter()
+        .filter(|item| conn_catalog.get_item(&item.id).item_type() == CatalogItemType::Source)
+        .map(|item| (item.id, item))
+        .collect();
+
+    // Iterate over all sources in the catalog.
+    for source in conn_catalog
+        .get_items()
+        .iter()
+        .filter(|item| item.item_type() == CatalogItemType::Source)
+    {
```
1. We [open a transaction](https://github.com/MaterializeInc/materialize/blob/5a25485f375c696347f0ad8f25b00461f27c98aa/src/adapter/src/catalog/open.rs#L234) for migration.
2. This uses the [`DurableCatalogState` impl for `PersistCatalogState`](https://github.com/MaterializeInc/materialize/blob/5a25485f375c696347f0ad8f25b00461f27c98aa/src/catalog/src/durable/persist.rs#L1290-L1293), which opens the txn [with a snapshot](https://github.com/MaterializeInc/materialize/blob/main/src/catalog/src/durable/persist.rs#L506).
3. This loads [`Snapshot`'s `items` field](https://github.com/MaterializeInc/materialize/blob/main/src/catalog/src/durable/objects.rs#L768) with all of the [`StateUpdateKind::Item`](https://github.com/MaterializeInc/materialize/blob/main/src/catalog/src/durable/persist.rs#L572-L574) values in the catalog's "trace" (`Vec<(StateUpdateKind, Timestamp, i64)>`), which is the [latest snapshot of the entire catalog](https://github.com/MaterializeInc/materialize/blob/main/src/catalog/src/durable/persist.rs#L213-L216).

We can trace this back further, but I take it on blind faith that this snapshot succeeds and contains all of the data from when we booted `envd`.

Another consideration here is that the upgrade tests succeed with this PR. If the txn didn't contain all of the catalog items, we wouldn't migrate the items in the catalog. This, then, would try to render the sources using the previous subsource style of rendering, which we've deprecated entirely, so something would likely have panicked. This isn't "proof" by any means, but helped me build some intuition.

sploiselle commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -155,13 +156,482 @@ pub(crate) async fn migrate(
+        .filter(|id| id_dedup.insert(*id))
+        .collect();
+
+    assign_new_user_global_ids(tx, conn_catalog, now, needs_new_id)
```
> all of an item's dependencies have `GlobalId`s less than its own.

This is wrong--I should have typed "dependents have `GlobalId`s less than its own" or "dependencies have `GlobalIds greater than its own."

I regret the typo!

rjobanp replied:
> This is wrong--I should have typed "dependents have GlobalIds less than its own" or "dependencies have `GlobalIds greater than its own."

I think this is backwards too though.

In `src/repr/src/global_id.rs` the docstring says 
> (items with greater `GlobalId`s can depend on those lesser but never the other way around)
which makes sense to me, since we'll create an object first with a lower id, then later another object will be created with a higher id that can depend on the first object.

rjobanp replied:
I'm also a little concerned that we don't have a test that covers this or proves the global-id-reordering is successful -- is that something we think the upgrade-tests should cover or do we need to add an explicit check?

sploiselle replied:
hah, ok––the original comment was right:

> all of an item's dependents have `GlobalId`s greater than its own.

This line from your comment confuses me:

> and then it looks like we've made needs_new_id ordered such that the IDs at the front of the vec need to be converted to IDs greater than those after them.

The items at the front of the vec are those that should have the smallest `GlobalId`s (they are the dependencies); those at the back should have greater `GlobalId`s (they are the dependents).

> I'm also a little concerned that we don't have a test that covers this or proves the global-id-reordering is successful -- is that something we think the upgrade-tests should cover or do we need to add an explicit check?

The upgrade tests cover this. Subsources would panic if their primary source didn't exist when they got created. There are upgrade tests that create PG sources and then upgrade them:

```
bin/mzcompose --dev --find platform-checks run default --scenario=UpgradeEntireMz --check=PgCdc 
```

Sadly this isn't working for me locally atm ([Slack](https://materializeinc.slack.com/archives/C01LKF361MZ/p1713972699539909)) but it is running in CI.

sploiselle replied:
Issue with running that commands was fortunately just a local issue.

rjobanp replied:
Gotcha, my comment came from what I think is happening with the `needs_new_id` vec above:

First we drain the vec into a double-ended queue, such that the queue starts with the ids of the subsources that currently have a `GlobalId` less than their primary source.

Then we iterate over this queue from the front, pushing each value we encounter into the `needs_new_id` vec. We add the dependents of each subsource to the end of the queue because we want these dependents to have higher global ids than the things they depend on (the subsource, to start).

So we end up with a `needs_new_id` vec in an order such that each subsequent element needs to have an id greater than the one before it.
Then we reverse this `needs_new_id` vec and filter out duplicate ids, such that the elements at the end must have ids less than those before them.

But this is opposite to the docstring on the `assign_new_user_global_ids` method, which says that the first element will have the smallest ID and the last element will have the greatest.

Let's sync live but I just wanted to write this out to make sure I could clarify

sploiselle replied:
I found the issue that was letting my upgrade tests pass with this bug in place. To resolve the issue, I've added asserts that all of an item's dependents have `GlobalId`s greater than itself, as well as adjusting the `PgCdc` platform test to be harder to fool in the upgrade scenario (added a new view with complex dependencies).

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -257,11 +261,146 @@ impl Coordinator {
+        let mut create_source_plans = Vec::with_capacity(subsource_stmts.len() + 2);
+
+        // 1. First plan the progress subsource
+        // Progress subsource must not have a dependency on the source.
```
This is why I dislike comments––all they do is get people asking hard questions. Here's my amendment:

```
// 1. First plan the progress subsource.
//
// The primary source depends on this subsource because the primary
// source needs to know its shard ID, and the easiest way of
// guaranteeing that the shard ID is discoverable is to create this
// collection first.
```

sploiselle commented on src/sql/src/plan/statement/acl.rs:
```diff
@@ -215,7 +215,7 @@ fn plan_alter_item_owner(
             }
 
             // Sub-sources cannot be altered directly.
```
ty!

sploiselle commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -835,7 +846,16 @@ impl Coordinator {
     }
 
-    fn drop_materialized_views(&mut self, mviews: Vec<(ClusterId, GlobalId)>) {
+    /// A convenience method for dropping sources.
```
ty!

sploiselle commented:
@rjobanp I've addressed the feedback locally and it'll be in the next push. Just avoiding unnecessary CI cycles.

rjobanp commented on src/storage-controller/src/lib.rs:
```diff
@@ -1197,41 +1130,89 @@ where
+        }
+
+        // Do not bother re-executing collections we know we plan to drop.
+        to_execute.retain(|id| !to_drop.contains(id));
```
we never push anything into `to_execute`. What is going on here?

sploiselle replied:
😮‍💨 this got lost in a rebase; if we remove a source export, we should re-execute its ingestion.

rjobanp reviewed (commented):
I think i've looked through almost all of the rust changes at this point. Planning to scan through the test changes tomorrow

guswynn reviewed (approved):
I have lots of inline questions/comment suggestions, but none are blocking!

I'm happy with the structure laid out here, but I want to note that there are 3 places that make me slightly nervous:
- The migration code, which @petrosagg and @rjobanp have looked at.
- The read hold code, which I only briefly looked at.
- The `plan_subsource` and `of_source` handling in the sequencer.

The latter two should be covered by the passing CI, but we may want to check in on them as we do the followup `TODO`'s

guswynn commented:
@rjobanp I looked through all the test changes, and they were straightforward except for the ones I commented on!

sploiselle commented on test/sqllogictest/object_ownership.slt:
```diff
@@ -2338,8 +2338,10 @@ ALTER CLUSTER materialize_public_s OWNER TO joe
 ALTER CLUSTER REPLICA materialize_public_s.linked OWNER TO joe
 
-statement error cannot ALTER this type of source
+simple conn=mz_system,user=mz_system
 ALTER SOURCE accounts OWNER TO joe
+----
+COMPLETE 0
```
ALTER...OWNER seems lightly documented: https://materialize.com/docs/sql/alter-owner/

@jkosh44 Do you see an RBAC issue with sources and subsources being able to have disparate owners? Maybe they need to be aligned in the way that tables and indexes are.

jkosh44 replied:
Being an owner of an object allows you to `DROP`, `ALTER`, `COMMENT`, `GRANT`/`REVOKE` that object. Note that running `DROP SOURCE s CASCADE`, only requires ownership of `s`, not any dependents of `s`. So I don't see any issue with sources and subsources being able to have disparate owners.

sploiselle replied:
ty! @guswynn will also add a note to the release notes

guswynn replied:
thanks!

sploiselle commented on src/storage-types/src/sources/postgres.rs:
```diff
@@ -310,11 +334,21 @@ impl AlterCompatible for PostgresSourcePublicationDetails {
+                match (timeline_id, &other.timeline_id) {
+                    (Some(curr_id), Some(new_id)) => curr_id == new_id,
+                    (None, Some(_)) => true,
+                    // New values must always have timeline ID
+                    (_, None) => false,
```
Just something I noticed was wrong.

sploiselle commented on src/storage-types/src/sources/mysql.rs:
```diff
@@ -249,6 +253,24 @@ pub struct MySqlSourceDetails {
+impl MySqlSourceDetails {
+    pub fn output_idx_for_name(&self, name: &UnresolvedItemName) -> Option<usize> {
+        let (schema_name, name) = match &name.0[..] {
+            // TODO(#26772): remove this inserted database name that doesn't refer to
+            // anything meaningful.
+            [database, schema_name, name] if database.as_str() == "mysql" => {
```
This is the _upstream_ reference. We insert the `mysql` string as the database because of how subsources used to resolve names.

guswynn replied:
ah i see!

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -87,6 +88,16 @@ pub enum IntrospectionType {
+    /// specifically the output identified by `external_reference`.
+    IngestionExport {
+        ingestion_id: GlobalId,
+        // This is an `UnresolvedItemName` because the structure works for PG,
+        // MySQL, and load generator sources. However, in the future, it should
+        // be sufficiently genericized to support all multi-output sources we
+        // support.
```
I don't think this is an actual `TODO` because we have no plan at the moment to actually do anything with this...ever AFAIK. The comment is just a very preemptive comment to a future engineer going, "Why does this need to be this way."

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -397,20 +379,25 @@ pub trait StorageController: Debug {
+    fn check_alter_ingestion_source_desc(
         &mut self,
-        collections: &BTreeMap<GlobalId, IngestionDescription>,
+        collections: &BTreeMap<GlobalId, SourceDesc>,
     ) -> Result<(), StorageError<Self::Timestamp>>;
 
-    /// Alter the identified collection to use the described ingestion.
-    async fn alter_collection(
+    /// Alters each identified collection to use the correlated [`SourceDesc`].
+    async fn alter_ingestion_source_desc(
```
`check_alter_ingestion_source_desc` gets called separately from `alter_ingestion_source_desc` to avoid panicks if we've borked something.

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -657,298 +732,156 @@ where
-            // Assess dependency since, which we have to fast-forward this
-            // collection's since to.
-            let storage_dependencies = collection.description.get_storage_dependencies();
+            // This circuitous pattern is because of lifetime rules––we cannot
+            // both borrow a mutable reference to the parent collection and
+            // immutable references to the exports.
```
That would be nice!

sploiselle commented on src/audit-log/src/lib.rs:
```diff
@@ -172,6 +174,8 @@ pub enum EventDetails {
     UpdateItemV1(UpdateItemV1),
     RenameSchemaV1(RenameSchemaV1),
     AlterRetainHistoryV1(AlterRetainHistoryV1),
+    ToNewIdV1(ToNewIdV1),
+    FromPreviousIdV1(FromPreviousIdV1),
```
From a billing perspective, this should be an atomic change. I don't think there is special billing that needs to be aware of specific `GlobalId`s. However, console needed to know about this.

sploiselle commented on src/catalog/src/durable/upgrade/snapshots/objects_v56.txt:
```diff
@@ -0,0 +1,100 @@
+CkoKSLoBRQoJCgNrZXkSAggEChoKBGtpbmQSEkIQU3lzdGVtUHJpdmlsZWdlcwocCgV2YWx1ZRITugEQCg4KCGFjbF9tb2RlEgIIBA==
```
¯\\\_(ツ)\_/¯

guswynn replied:
I feel you on that!

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -1127,6 +1134,8 @@ impl_display_for_with_option!(CreateSubsourceOption);
 pub struct CreateSubsourceStatement<T: AstInfo> {
     pub name: UnresolvedItemName,
     pub columns: Vec<ColumnDef<T>>,
+    // Tracks the primary source of this subsource if a data-bearing subsource.
+    pub of_source: Option<T::ItemName>,
```
There are two types of subsources––data bearing and the progress subsources. In my mind, the progress collection isn't "data bearing" because it's just metadata...but this term could use some workshopping.

sploiselle commented on src/adapter/src/optimize/dataflows.rs:
```diff
@@ -296,7 +296,8 @@ impl<'a> DataflowBuilder<'a> {
         // we can retrieve monotonicity information from the parent source.
         match &source.data_source {
             DataSourceDesc::Ingestion(ingestion) => ingestion.desc.monotonic(),
-            DataSourceDesc::Introspection(_)
+            DataSourceDesc::IngestionExport { .. }
+            | DataSourceDesc::Introspection(_)
```
Maybe you mised `mz_storage_types::sources::load_generator::LoadGenerator::is_monotonic`?

guswynn replied:
If a compute dataflow reads from the subsource shard though, it will hit this case, right?

sploiselle replied:
Ohhhh, yeah, totally. Sorry I misunderstood. https://github.com/MaterializeInc/materialize/issues/26808

sploiselle commented on src/sql/src/pure/postgres.rs:
```diff
@@ -155,24 +159,25 @@ pub(super) fn generate_text_columns(
     (
-        Vec<CreateSourceSubsource<Aug>>,
-        Vec<(GlobalId, CreateSubsourceStatement<Aug>)>,
+        Vec<CreateSubsourceStatement<Aug>>,
+        // These are the tables referenced by the subsources. We want this set
+        // of tables separately so we can retain only table definitions that are
+        // referenced by subsources. This helps avoid issues when generating PG
+        // source table casts.
+        Vec<PostgresTableDesc>,
```
You are correct, however both callers of this fn immediately want to split the values, so that seems annoying. The places that they're used that assert that the vecs are of the same length are also very far away.

sploiselle commented on src/sql/src/pure.rs:
```diff
@@ -1185,55 +1189,43 @@ async fn purify_create_source(
                 let subsource = CreateSubsourceStatement {
                     name: subsource_name,
                     columns,
+                    // We don't know the primary source's `GlobalId` yet; fill
+                    // it in once we generate it.
+                    of_source: None,
```
The struct had a comment but not a doc comment because I can't reliably count to 3.

sploiselle replied:
Also cannot add a module reference because of dependency issues.

sploiselle commented on src/sql/src/pure.rs:
```diff
@@ -1185,55 +1189,43 @@ async fn purify_create_source(
+            // Now that we know which subsources to create alongside this
+            // statement, remove the references so it is not canonicalized as
+            // part of the `CREATE SOURCE` statement in the catalog.
+            *referenced_subsources = None;
```
This applies to all of them equally but I wouldn't expect someone to try and find the documentation for what this does in other similarly shaped parts of the code. Maybe I'm misunderstanding the question?

guswynn replied:
I was just saying that the code seems to do `*referenced_subsources = None;` twice in some cases, but its not important to change! I could also be reading the nesting wrong cause something something github

sploiselle commented on src/sql/src/plan/statement/show.rs:
```diff
@@ -909,20 +914,142 @@ fn humanize_sql_for_show_create(
-            }
-            CreateSourceConnection::MySql { options, .. } => {
-                options.retain(|o| o.name != MySqlConfigOptionName::Details)
+        // `CREATE SOURCE` statements should roundtrip. However, sources and
+        // their subsources have a complex relationship, so we need to do a lot
+        // of work to reconstruct the statement for multi-output sources.
```
Yep! In my defense, I have similar comments inline but this is good, too.

sploiselle commented on src/sql/src/plan/statement/show.rs:
```diff
@@ -909,20 +914,142 @@ fn humanize_sql_for_show_create(
+                })
+                .collect();
+
+            tracing::info!("curr_references {:?}", curr_references);
```
le sigh--meant to just remove these entirely

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -3158,184 +3127,88 @@ where
+    ///
+    /// Note that this might include fast-forwarding new collections' sinces as
+    /// far as their dependencies' sinces.
+    fn determine_collection_dependencies(&mut self, id: GlobalId) -> Result<(), StorageError<T>> {
```
I significantly refactored this area of the code such that the function's content is now appropriate to be called `install_dependency_read_holds`. Will ping you with the commit that could use another review.

sploiselle commented on src/storage/src/source/mysql.rs:
```diff
@@ -124,14 +125,16 @@ impl SourceRender for MySqlSourceConnection {
         let mut table_info = BTreeMap::new();
-        for (i, desc) in self.details.tables.iter().enumerate() {
+        for (_id, SourceExport { output_index, .. }) in &config.source_exports {
+            // Output index 0 is the primary source which is not a table.
```
Added documentation to the field.

sploiselle commented on test/legacy-upgrade/check-from-v0.45.0-kafka-progress.td:
```diff
@@ -19,4 +19,4 @@ $ set-regex match=testdrive-upgrade-kafka-source-.*?' replacement=<TOPIC>'
 materialize.public.kafka_source "CREATE SOURCE \"materialize\".\"public\".\"kafka_source\" FROM KAFKA CONNECTION \"materialize\".\"public\".\"kafka_conn\" (TOPIC = '<TOPIC>') FORMAT AVRO USING SCHEMA '{   \"type\": \"record\",   \"name\": \"cpx\",   \"fields\": [     {\"name\": \"a\", \"type\": \"long\"},     {\"name\": \"b\", \"type\": \"long\"}   ] }' ENVELOPE NONE EXPOSE PROGRESS AS \"material ...
 
-> DROP SOURCE kafka_source;
+> DROP SOURCE kafka_source CASCADE;
```
We shouldn't! Will remove this.

sploiselle commented on src/storage-client/src/controller.rs:
```diff
@@ -87,6 +88,16 @@ pub enum IntrospectionType {
 pub enum DataSource {
     /// Ingest data from some external source.
     Ingestion(IngestionDescription),
+    /// This source receives its data from the identified ingestion,
+    /// specifically the output identified by `external_reference`.
+    IngestionExport {
```
ty

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -569,61 +578,127 @@ where
+        // Reorder in dependency order.
+        to_register.sort_by_key(|(id, ..)| *id);
+
+        let mut to_execute = BTreeSet::new();
+        let mut new_collections = BTreeSet::new();
```
done

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -569,61 +578,127 @@ where
-                    vec![],
-                    metadata.clone(),
-                );
+        // Statistics need a level of indirection so we can mutably borrow
+        // `self` when registering collections and when we are inserting
+        // statistics.
```
:D

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -569,61 +578,127 @@ where
+
+                    new_source_statistic_entries.insert(id);
+                }
+                DataSource::Other(DataSourceOther::TableWrites) => {
+                    debug!(desc = ?description, meta = ?metadata, "registering {} with persist table worker", id);
+                    table_registers.push((id, write, collection_state));
+                }
+                DataSource::Progress | DataSource::Other(DataSourceOther::Compute) => {
+                    debug!(desc = ?description, meta = ?metadata, "not registering {} with a controller persist worker", id);
+                    self.collections.insert(id, collection_state);
+                }
+                DataSource::Ingestion(_) => {
+                    debug!(desc = ?description, meta = ?metadata, "not registering {} with a controller persist worker", id);
+                    self.collections.insert(id, collection_state);
+                    new_source_statistic_entries.insert(id);
+                }
```
Oy––this was a good comment. I think the original thing I did for `execute_collections` was too easy to misuse (the only thing that should ever be re-executed are ingestions). I refactored this to `run_ingestion` and it now only runs ingestions (since that is the only collection type that should ever be re-rendered).

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -3545,6 +3612,10 @@ impl<T: Timestamp> CollectionState<T> {
+            | DataSource::IngestionExport { .. }
             | DataSource::Progress => None,
         }
     }
```
Such a good call to investigate moving things onto `DataSource`. I'm going to add that as a `TODO` just to minimize the chaos here. I've also got a doc comment staged locally that will be visible in the next push.

guswynn replied:
Yes 100% just make it a TODO!

sploiselle commented on src/catalog/src/memory/objects.rs:
```diff
@@ -617,12 +635,27 @@ impl Source {
+                    // source's data shard, so we don't include it in accounting
+                    // for users.
+                    GenericSourceConnection::Postgres(_) | GenericSourceConnection::MySql(_) => 0,
+                    GenericSourceConnection::LoadGenerator(lg) => {
+                        if lg.load_generator.views().is_empty() {
+                            // Load generator writes directly to its persist shard
+                            1
+                        } else {
+                            // Load generator has 1 persist shard per output,
+                            // which will be accounted for by `SourceExport`.
+                            0
+                        }
```
I'm going to punt on this for the moment but will add a TODO.

sploiselle commented on src/adapter/src/coord.rs:
```diff
@@ -1473,113 +1470,7 @@ impl Coordinator {
         debug!("coordinator init: initializing storage collections");
         self.bootstrap_storage_collections().await;
 
-        // Load catalog entries based on topological dependency sorting. We do
```
An understatement!

sploiselle commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -4132,10 +4140,10 @@ impl Coordinator {
                 ops.extend(dependent_index_ops);
 
-                // Alter owner cascades down to sub-sources and progress collections.
+                // Alter owner cascades down to progress collections.
```
pinged Joe about this

sploiselle replied:
Added release note

sploiselle commented:
@rjobanp Here's a commit resolving your comment/request to clarify the `CollectionState` dependencies and the fn name for installing collection dependencies + read holds. https://github.com/MaterializeInc/materialize/pull/26556/commits/16b1b2c6875ae3a1ea2d441e8e68e4b879b8bf8d

The rest of the fixes have been resolved inline in the appropriate commits.

rjobanp reviewed (approved):
cool just one more comment and then as long as other comments have all been addressed this LGTM

sploiselle commented on src/storage-controller/src/lib.rs:
```diff
@@ -569,61 +587,179 @@ where
+            // Determine the intial since of the collection.
+            let initial_since = match storage_dependency {
+                Some(dep) => {
+                    let remap_collection = self.collection(dep)?;
```
Oh that is just a poorly named variable––I meant to just call it `dependency_collection`. Easy fix.

sploiselle commented:
@RobinClowers This change is ready to merge––any extra coordination we need on the frontend?

RobinClowers commented:
@sploiselle Great! I have a draft PR that handles it, so I'll just need to update it to target the correct release. My understanding is that we will cut v0.98.0 next week, and presumably this will be in it? Feel free to merge, I'll make sure to get my PR merged early next week so we are ready for it.
